src/Workspaces/Core/Portable/Remote/RemoteHostClient.cs

```diff
@@ -104,35 +62,57 @@ private void OnConnectionChanged(bool connected)
             ConnectionChanged?.Invoke(this, connected);
         }
 
-        private async Task<PinnedRemotableDataScope> GetPinnedScopeAsync(Solution solution, CancellationToken cancellationToken)
+        /// <summary>
+        /// NoOpClient is used if a user killed our remote host process. Basically this client never
+        /// create a session
+        /// </summary>
+        public class NoOpClient : RemoteHostClient
         {
-            if (solution == null)
+            public NoOpClient(Workspace workspace) :
+                base(workspace)
+            {
+            }
+
+            public override Task<Session> TryCreateSessionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
             {
-                return null;
+                return SpecializedTasks.Default<Session>();
             }
 
-            Contract.ThrowIfFalse(solution.Workspace == _workspace);
+            protected override void OnConnected()
+            {
+                // do nothing
+            }
 
-            var service = _workspace.Services.GetService<ISolutionSynchronizationService>();
-            return await service.CreatePinnedRemotableDataScopeAsync(solution, cancellationToken).ConfigureAwait(false);
+            protected override void OnDisconnected()
+            {
+                // do nothing
+            }
         }
 
-        // TODO: make this to not exposed to caller. abstract all of these under Request and Response mechanism
         public abstract class Session : IDisposable
         {
-            protected readonly PinnedRemotableDataScope PinnedScopeOpt;
             protected readonly CancellationToken CancellationToken;
+            private PinnedRemotableDataScope _scope;
 
             private bool _disposed;
 
-            protected Session(PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+            protected Session(CancellationToken cancellationToken)
             {
                 _disposed = false;
+                _scope = null;
 
-                PinnedScopeOpt = scope;
                 CancellationToken = cancellationToken;
             }
 
+            public PinnedRemotableDataScope PinnedRemotableDataScope => _scope;
```

Comment:
I want Session to be simple session that each type of remote host such as Service Hub provide implementation which just know how to make connection and communicate each other. 

and I want something other to tie OOP layer to Roslyn concept such as Solution.

so that I can optimize OOP such as pooling session without constrained by Roslyn's restriction. that should be done by the other thing that tie OOP to Roslyn.

also, that will let people to have different kind of connection such as one that tie to solution, one that kept alive. one that can be used with multiple solution and etc.

people who consume OOP doesn't even need to this. they just need to use helper type that suite to their need. 

or they can use bare bone session if they want to build completely their own behavior. for example, this monopolistic approve caused me to break things every single time I need new behavior because underlying system baked in all things tightly coupled together and removed all the flexibility. first time, I needed session that doesn't require solution, I had to hack that in. first time I needed a session with multiple solution, I had to hack that in. first time I need session that lives long that support restart I had to again hack that in. I want all the building block to be there and other monolithic stuff to be helpers.